File: solver.py
import sys
import logging
import time


def getFreq(char, word):
    count = 0
    for letter in word:
        if char == letter:
            count += 1
    return count


def answer_generator(user_word, correctness):
    global ALL_WORDS
    for i in range(5):
        log.debug("Looking at letter:", user_word[i])
        for word in ALL_WORDS.keys():
            if correctness[i] == '_' and user_word[i] in word:
                log.debug("Removing: %s invalid letter: %s", word,
                          user_word[i])
                ALL_WORDS[word] = False
                for j in range(5):
                    if i != j and user_word[i] == user_word[j]:
                        if correctness[j] == '!':
                            ALL_WORDS[word] = word[j] == user_word[i] and (user_word[i] not in word[j+1:] and user_word[i] not in word[:j])
                            continue
                        elif correctness[j] == '?':
                            ALL_WORDS[word] = word[j] != user_word[i] and (user_word[i] in word[i+1:] or user_word[i] in word[:i])
                            continue
            elif correctness[i] == '!' and user_word[i] != word[i]:
                log.debug("Removing: %s missing correct letter: %s", word,
                          user_word[i])
                ALL_WORDS[word] = False
                continue
            elif correctness[i] == '?':
                if user_word[i] not in word or user_word[i] == word[i]:
                    log.debug("Removing: %s missing correct letter or in wrong postition: %s",
                              word, user_word[i])
                    ALL_WORDS[word] = False
                    continue
                else:
                    for j in range(5):
                        if i != j and user_word[i] == user_word[j]:
                            if correctness[j] == '!':
                                ALL_WORDS[word] = (word[i] != user_word[i] and word[j] == user_word[i]) and (user_word[i] in word[j+1:] or user_word[i] in word[:j])
                                continue
                            if correctness[j] == '?':
                                word_freq = getFreq(user_word[i], word)
                                user_freq = getFreq(user_word[i], user_word)
                                ALL_WORDS[word] = word[i] != user_word[i] and word[j] != user_word[i] and (word_freq == user_freq)
                                continue


def main():
    global ALL_WORDS
    guesses = 0
    print(len(sys.argv))
    if len(sys.argv) > 1 and str.lower(sys.argv[1]) == "debug":
        log.setLevel(logging.DEBUG)
    with open('words.txt') as f:
        ALL_WORDS = {w.strip(): True for w in f.readlines()}
    repeat = 'y'
    while (repeat != 'n'):
        print('You have {:d} guesses remaining'.format(6 - guesses))
        user_word = input("Enter your 5 letter word: ").strip().lower()
        correctness = input("Enter _ for not in word, ! for correct spot and ? if in wrong position: ").strip()
        if (correctness == "!!!!!"):
            print("The word is:", user_word + "!")
            exit(0)

        #start = time.time()
        answer_generator(user_word, correctness)
        #end = time.time()

        #diff = (end - start) * 1000
        #print("Execution time:", diff, "milliseconds")

        ALL_WORDS = dict(filter(lambda entry: entry[1], ALL_WORDS.items()))
        valid_word_count = len(ALL_WORDS.keys())
        if valid_word_count > 1:
            print("Valid Word(s): ", end="")
            for k in ALL_WORDS.keys():
                print(k, end=" ")
            print("")
        elif valid_word_count == 1:
            print("The answer is:", next(iter(ALL_WORDS.keys())) + "!")
            exit(0)
        else:
            print("No valid words, somethings wrong!!!")
            exit(1)
        repeat = input("Guess another word? y/n: ").strip()
        guesses += 1
        print("")


log = logging.getLogger('solver')
log.setLevel(logging.INFO)
handler = logging.StreamHandler(sys.stdout)
dt_fmt = '%Y-%m-%d %H:%M:%S'
formatter = logging.Formatter('[%(asctime)s.%(msecs)03d] [%(levelname)s] %(name)s: %(message)s', dt_fmt)
handler.setFormatter(formatter)
log.addHandler(handler)
ALL_WORDS = dict()
main()

File: solver.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"sort"
	"strings"
)

var all_words map[string] bool = make(map[string]bool)
var debug_logger *log.Logger

func get_freq(character string, word string) int {
    counter := 0
    for _, char := range word {
        if string(char) == character {
            counter++
        }
    }
    return counter
}

func load_word_list() {
    file, err := os.Open("words.txt")

    if err != nil {
        log.Fatal(err)
    }

    defer file.Close()

    scanner := bufio.NewScanner(file)

    for scanner.Scan() {
        all_words[scanner.Text()] = true
    }

    if err := scanner.Err(); err != nil {
        log.Fatal(err)
    }
}

func filter(word_map map[string] bool) map[string] bool {
    filtered_word_map := make(map[string] bool)
    for k, v := range word_map {
        if v { filtered_word_map[k] = v }
    }
    return filtered_word_map

}

func answer_generator(user_word string, correctness string) {
    for i := range 5 {
        checking_char := string(user_word[i])
        debug_logger.Printf("Loooking at letter %s\n", checking_char)
        for word := range all_words {
            if string(correctness[i]) == "_" && strings.Contains(word, checking_char) {
                debug_logger.Printf("Removing %s invalid letter: %s\n", word, checking_char)
                all_words[word] = false
                for j := range 5 {
                    correctness_char := string(correctness[j])
                    if i != j && checking_char == string(user_word[j]) {
                        if correctness_char == "!" {
                            all_words[word] = string(word[j]) == checking_char && (!strings.Contains(word[j+1:], checking_char) && !strings.Contains(word[:j], checking_char))
                            debug_logger.Println(word, "is:", all_words[word])
                            continue
                        } else if correctness_char == "?" {
                            all_words[word] = string(word[j]) != checking_char && (strings.Contains(word[i+1:], checking_char) || strings.Contains(word[:i], checking_char))
                            debug_logger.Println(word, "is:", all_words[word])
                            continue
                        }
                    }
                }
            } else if string(correctness[i]) == "!" && checking_char != string(word[i]) {
                debug_logger.Printf("Removing %s missing correct letter: %s\n", word, checking_char)
                all_words[word] = false
                continue
            } else if string(correctness[i]) == "?" { // ? case
                if !strings.Contains(word, checking_char) || checking_char == string(word[i]) {
                    debug_logger.Printf("Removing: %s missing correct letter or in wrong postition: %s\n", word, checking_char)
                    all_words[word] = false
                    continue
                } else {
                    var correctness_char string
                    for j := range 5 {
                        correctness_char = string(correctness[j])
                        if i != j && checking_char == string(user_word[j]) {
                            if correctness_char == "!" {
                                debug_logger.Printf("split char: %s, contains true? %t", word[:j], strings.Contains(word[:j], checking_char))
                                all_words[word] = (string(word[i]) != checking_char && string(word[j]) == checking_char) && (strings.Contains(word[j+1:], checking_char) || strings.Contains(word[:j], checking_char))
                                debug_logger.Println(word, "is:", all_words[word])
                                continue
                            } else if correctness_char == "?" {
                                word_freq := get_freq(checking_char, word)
                                user_freq := get_freq(checking_char, user_word)
                                all_words[word] = string(word[i]) != checking_char && string(word[j]) != checking_char && (word_freq == user_freq)
                                debug_logger.Println(word, "is:", all_words[word])
                                continue
                            }
                        }
                    }
                }
            }
        }
    }
}

func main() {
    reader := bufio.NewReader(os.Stdin)
    guesses := 0
    var argv []string = os.Args
    if len(argv) > 1 && strings.ToLower(argv[1]) == "debug" {
        debug_logger = log.New(os.Stdout, "DEBUG: ", log.Ldate|log.Ltime|log.Lshortfile)
    } else {
        debug_file, err := os.OpenFile("go-logs.log", os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0666)
        if err != nil {
            log.Fatal(err)
        }
        debug_logger = log.New(debug_file, "DEBUG: ", log.Ldate|log.Ltime|log.Lshortfile)
    }

    load_word_list()
    //for k := range all_words {
    //    delete(all_words, k)
    //}
    //all_words["arles"] = true

    repeat := "y"
    for guesses <= 6 && repeat != "n" {
        fmt.Printf("You have %d guesses remaining\n", 6 - guesses)
        fmt.Printf("Enter your 5 letter word: ")
        user_word, _ := reader.ReadString('\n')
        cleaned_user_word := strings.ToLower(strings.TrimSpace(user_word))
        fmt.Printf("Enter _ for not in word, ! for correct spot and ? if in wrong position: ")
        correctness, _ := reader.ReadString('\n')
        cleaned_correctness := strings.ToLower(strings.TrimSpace(correctness))
        if cleaned_correctness == "!!!!!" {
            fmt.Println("The word is:", cleaned_user_word, "!")
            os.Exit(0)
        }

        answer_generator(cleaned_user_word, cleaned_correctness)

        all_words = filter(all_words) 
        valid_word_count := len(all_words)
        if valid_word_count > 1 {
            keys := make([]string, 0, len(all_words))
            for k := range all_words {
                keys = append(keys, k)
            }
            sort.Strings(keys)
            fmt.Printf("Valid Word(s): ")
            for _, k := range keys {
                fmt.Printf("%s ", k)
            }
            fmt.Println()
        } else if valid_word_count == 1 {
            var answer string
            for k := range all_words { answer = k; break }
            fmt.Printf("The answer is: %s\n", answer)
            os.Exit(0)
        } else {
            fmt.Println("No valid words, something is wrong!!!")
            os.Exit(1)
        }
        fmt.Printf("Guess another word? y/n: ")
        repeat_read, _ := reader.ReadString('\n')
        repeat = strings.TrimSpace(repeat_read)
        guesses++ 
        fmt.Println()
    }

}
